### PR TITLE
Manipulation: Respect script crossorigin attribute in DOM manipulation

### PR DESCRIPTION
**Pull request by gaohuia**
### Summary ###
jQuery does not respect the crossOrigin attribute on the script tag #4542

### Checklist ###

* [x] All authors have signed the CLA at https://cla.js.foundation/jquery/jquery
* [x] New tests have been added to show the fix or feature works
* [x] Grunt build and unit tests pass locally with these changes
* [ ] If needed, a docs issue/PR was created at https://github.com/jquery/api.jquery.com

### CODE REVIEW

**Comment by mgol** [comment]
Thanks for the PR, I'll have a closer look later. In the meantime, please sign our CLA, this is a requirement for us moving forwards with the PR.

File: src/ajax/script.js [hunk 1]
```diff
@@ -46,6 +46,7 @@ jQuery.ajaxTransport( "script", function( s ) {
 		return {
 			send: function( _, complete ) {
 				script = jQuery( "<script>" )
+					.attr( "crossOrigin", s.crossOrigin || null )
 					.attr( s.scriptAttrs || {} )
```
Comment (jaubourg) [inline src/ajax/script.js]: That's what the `scriptAttrs` option if for. You can add any attribute in there with no need to introduce a new option. It was designed specifically to avoid option-creep.

The correct route would be to add `crossOrigin` into the `scriptAttrs` object.

Comment (gaohuia) [reply src/ajax/script.js]: Updated.

File: src/manipulation.js [hunk 2]
```diff
@@ -162,7 +162,9 @@ function domManip( collection, args, callback, ignored ) {
 							// Optional AJAX dependency, but won't run scripts if not present
 							if ( jQuery._evalUrl && !node.noModule ) {
 								jQuery._evalUrl( node.src, {
-									nonce: node.nonce || node.getAttribute( "nonce" )
+									nonce: node.nonce || node.getAttribute( "nonce" ),
+									crossOrigin: node.crossOrigin ||
```
Comment (mgol) [inline src/manipulation.js]: I don't think you need to read both the attribute & the property. Support for both seems similar:
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script
https://developer.mozilla.org/en-US/docs/Web/API/HTMLScriptElement
(look at the "Browser compatibility" tables below).

The reason why for `nonce` we needed both the property & the attribute is that in Chrome (& latest standards) `getAttribute( "nonce" )` doesn't return an up to date `nonce` value:
https://github.com/whatwg/html/issues/2369
https://html.spec.whatwg.org/#nonce-attributes
but, at the same time, the `nonce` property is not supported in Firefox & Edge so we need the `getAttribute` fallback for them.

I'd just leave `crossOrigin: node.crossOrigin` here. But please make sure the tests still pass.

**Comment by gaohuia** [comment]
Thanks for your explaination.

File: test/unit/core.js [hunk 3]
```diff
@@ -197,6 +197,34 @@ QUnit.test( "globalEval execution after script injection (#7862)", function( ass
...
+	var done = assert.async();
+	assert.expect( 1 );
+
+	var timeout = null;
+
+	Globals.register( "corsCallback" );
+	window.corsCallback = function( response ) {
+		assert.ok( response === true, "Cors" );
+
+		if ( timeout != null ) {
```
Comment (mgol) [inline test/unit/core.js]: The `if` is not necessary, `clearTimeout` on `null` is a no-op.

File: test/unit/core.js [hunk 4]
```diff
@@ -197,6 +197,34 @@ QUnit.test( "globalEval execution after script injection (#7862)", function( ass
...
+			window.clearTimeout( timeout );
+		}
+
+		done();
+	};
+
+	var src = baseURL + "mock.php?action=cors";
+	src = src.replace( "localhost", "127.0.0.1" );
```
Comment (mgol) [inline test/unit/core.js]: Is this simulating a cross-origin request by fetching a script from `127.0.0.1` at a `localhost` host? If that's the case, please add a comment above this line (with an empty line preceeding the comment).

File: test/unit/core.js [hunk 5]
```diff
@@ -197,6 +197,34 @@ QUnit.test( "globalEval execution after script injection (#7862)", function( ass
...
+		assert.ok( response === true, "Cors" );
+
+		if ( timeout != null ) {
+			window.clearTimeout( timeout );
+		}
+
+		done();
+	};
+
+	var src = baseURL + "mock.php?action=cors";
```
Comment (mgol) [inline test/unit/core.js]: We typically put all var declarations near the top of the function under one `var` keyword, one declaration per line if immediately assigned, unassigned vars all in one line (the first one). Please see other tests for comparison.

Here I'd expect `assert.expect( 1 );` to be the first line, then an empty line & then all vars declared with one var statement.

File: test/unit/core.js [hunk 6]
```diff
@@ -197,6 +197,34 @@ QUnit.test( "globalEval execution after script injection (#7862)", function( ass
...
+	src = src.replace( "localhost", "127.0.0.1" );
+	var html = "<script type=\"text/javascript\" src=\"" + src + "\" crossOrigin=\"anonymous\"><\/script>";
+
+	jQuery( document.body ).append( html );
+	timeout = window.setTimeout( function() {
+		assert.ok( false, "Cors" );
```
Comment (mgol) [inline test/unit/core.js]: This assertion fires for me in Safari 13, Edge 18 & IE 11. It correctly doesn't fire only in Firefox 72 & Chrome 79.

We need to figure out why it fails in other browsers. For IE we might have to skip the test as it doesn't support the crossorigin attribute (by switching from `QUnit.test` to `QUnit.testUnlessIE`). Other browsers should pass it, though.

File: test/unit/core.js [hunk 7]
```diff
@@ -197,6 +197,34 @@ QUnit.test( "globalEval execution after script injection (#7862)", function( ass
...
+		if ( timeout != null ) {
+			window.clearTimeout( timeout );
+		}
+
+		done();
+	};
+
+	var src = baseURL + "mock.php?action=cors";
+	src = src.replace( "localhost", "127.0.0.1" );
+	var html = "<script type=\"text/javascript\" src=\"" + src + "\" crossOrigin=\"anonymous\"><\/script>";
```
Comment (mgol) [inline test/unit/core.js]: HTML is case-insensitive, so:
```suggestion
	var html = "<script type=\"text/javascript\" src=\"" + src + "\" crossorigin=\"anonymous\"><\/script>";
```

**Comment by mgol** [comment]
@gaohuia Your PR now contains my cherry-picked PRs... Can you instead rebase your PR over current master so that it only contains your code? My commits should disappear then:
```
git fetch --all --tags -p
git rebase origin/master
```
If your upstream is named differently than `origin`, substitute the name.

In the future, it's a good idea to not modify code on the `master` branch but keep `master` synced with upstream instead and use a different branch for PRs. See this link for more info:
https://contribute.jquery.org/commits-and-pull-requests/#never-commit-on-master

This is for the future sake only, let's keep this PR as it is as you can't change the source branch when the PR is open.

**Comment by mgol** [comment]
@gaohuia You can see that jQuery Slim tests failed on your PR (see the Travis link). Your unit test should be moved to `unit/ajax.js` where tests related to the ajax module should lie. Then that test will be excluded for the Slim build & it should pass again.

Also, remember about https://github.com/jquery/jquery/pull/4563#discussion_r364368782 - we still need to get the tests to succeed in IE, Edge & Safari.

**Comment by gaohuia** [comment]
Hi @mgol I didn't see any failed tests in safari 13, can you try again?

**Comment by mgol** [comment]
@gaohuia Yes, Safari 13 is still failing. Note that Travis only runs tests on Chrome & Firefox, you need to test on the other browsers manually.

See e.g. Safari 13:
Screen Shot 2020-01-17 at 23 03 56
and Edge 18:
Screen Shot 2020-01-17 at 23 09 56

**Comment by mgol** [comment]
Hmm, the above test results were achieved in our PHP setup, the one that runs on our TestSwarm as well: http://swarm.jquery.org/project/jquery

When I run it on the Karma setup, by running `grunt karma:chrome-debug`, opening http://localhost:9876 in Safari & clicking "DEBUG", all tests pass.

**Comment by mgol** [comment]
However, in Chrome & Firefox tests pass in both modes.

File: test/unit/ajax.js [hunk 8]
```diff
@@ -2840,6 +2840,32 @@ if ( typeof window.ArrayBuffer === "undefined" || typeof new XMLHttpRequest().re
 		} );
 	} );
 
+	QUnit.test( "jQuery.append with crossOrigin attribute", function( assert ) {
```
Comment (mgol) [inline test/unit/ajax.js]: The attribute is lowercase so:
```suggestion
	QUnit.test( "jQuery.append with crossorigin attribute", function( assert ) {
```

File: test/unit/ajax.js [hunk 9]
```diff
@@ -2840,6 +2840,32 @@ if ( typeof window.ArrayBuffer === "undefined" || typeof new XMLHttpRequest().re
...
+		Globals.register( "corsCallback" );
+		window.corsCallback = function( response ) {
+			assert.ok( response === true, "Cors" );
+			window.clearTimeout( timeout );
+			done();
+		};
+
+		var src = baseURL + "mock.php?action=cors";
+
+		// To simulate a cross-origin request
```
Comment (mgol) [inline test/unit/ajax.js]:
```suggestion
		// Simulate a cross-origin request
```

File: test/unit/ajax.js [hunk 10]
```diff
@@ -2840,6 +2840,32 @@ if ( typeof window.ArrayBuffer === "undefined" || typeof new XMLHttpRequest().re
 		} );
 	} );
 
+	QUnit.test( "jQuery.append with crossOrigin attribute", function( assert ) {
+		var done = assert.async();
```
Comment (mgol) [inline test/unit/ajax.js]: We typically put all var declarations near the top of the function under one `var` keyword, one declaration per line if immediately assigned, unassigned vars all in one line (the first one). Please see other tests for comparison.

Here I'd expect `assert.expect( 1 );` to be the first line, then an empty line & then all vars declared with one var statement.

File: test/unit/ajax.js [hunk 11]
```diff
@@ -2840,6 +2840,32 @@ if ( typeof window.ArrayBuffer === "undefined" || typeof new XMLHttpRequest().re
...
+
+		var src = baseURL + "mock.php?action=cors";
+
+		// To simulate a cross-origin request
+		src = src.replace( "localhost", "127.0.0.1" );
+		var html = "<script type=\"text/javascript\" src=\"" + src + "\" crossorigin=\"anonymous\"><\/script>";
+
+		jQuery( document.body ).append( html );
+		timeout = window.setTimeout( function() {
+			assert.ok( false, "Cors" );
```
Comment (mgol) [inline test/unit/ajax.js]: This assertion fires for me in Safari 13, Edge 18 & IE 11. It correctly doesn't fire only in Firefox 72 & Chrome 79.

We need to figure out why it fails in other browsers. For IE we might have to skip the test as it doesn't support the crossorigin attribute (by switching from `QUnit.test` to `QUnit.testUnlessIE`). Other browsers should pass it, though.

From my recent investigation: it seems the tests pass in Safari in Karma but not via the PHP setup. On the other hand, they succeed in both modes in Chrome & Firefox.

File: test/middleware-mockserver.js [hunk 12]
```diff
@@ -238,6 +238,15 @@ var mocks = {
...
 			resp.end( "QUnit.assert.ok( false, \"Mock return erroneously executed\" );" );
 		}
+	},
+	cors: function( req, resp ) {
+		if ( typeof req.headers[ "origin" ] != "undefined" ) {
+			resp.setHeader( "Access-Control-Allow-Methods", "GET" );
+			resp.setHeader( "Access-Control-Allow-Origin", req.headers[ "origin" ] );
+			resp.end( "corsCallback(true)" );
+		} else {
+			resp.end( "corsCallback(false)" );
```
Comment (mgol) [inline test/middleware-mockserver.js]:
```suggestion
			resp.end( "corsCallback( false )" );
```

File: test/middleware-mockserver.js [hunk 13]
```diff
@@ -238,6 +238,15 @@ var mocks = {
 		} else {
 			resp.end( "QUnit.assert.ok( false, \"Mock return erroneously executed\" );" );
 		}
+	},
+	cors: function( req, resp ) {
+		if ( typeof req.headers[ "origin" ] != "undefined" ) {
+			resp.setHeader( "Access-Control-Allow-Methods", "GET" );
+			resp.setHeader( "Access-Control-Allow-Origin", req.headers[ "origin" ] );
+			resp.end( "corsCallback(true)" );
```
Comment (mgol) [inline test/middleware-mockserver.js]:
```suggestion
			resp.end( "corsCallback( true )" );
```

File: test/data/mock.php [hunk 14]
```diff
@@ -229,6 +229,17 @@ protected function errorWithScript( $req ) {
 		}
 	}
 
+	protected function cors($req) {
+		if (isset($req->headers['ORIGIN'])) {
```
Comment (mgol) [inline test/data/mock.php]:
```suggestion
		if ( isset( $req->headers['ORIGIN'] ) ) {
```

File: test/data/mock.php [hunk 15]
```diff
@@ -229,6 +229,17 @@ protected function errorWithScript( $req ) {
...
+			$origin = $req->headers['ORIGIN'];
+			header("Access-Control-Allow-Methods: GET");
```
Comment (mgol) [inline test/data/mock.php]:
```suggestion
			header( "Access-Control-Allow-Methods: GET" );
```

File: test/data/mock.php [hunk 16]
```diff
@@ -229,6 +229,17 @@ protected function errorWithScript( $req ) {
...
+			header("Access-Control-Allow-Origin: " . $origin);
```
Comment (mgol) [inline test/data/mock.php]:
```suggestion
			header( "Access-Control-Allow-Origin: " . $origin );
```

File: test/data/mock.php [hunk 17]
```diff
@@ -229,6 +229,17 @@ protected function errorWithScript( $req ) {
...
+			corsCallback(true);
```
Comment (mgol) [inline test/data/mock.php]: There's a bug here, this will not actually do anything. It should be:
```suggestion
			echo 'corsCallback( true )';
```

File: test/data/mock.php [hunk 18]
```diff
@@ -229,6 +229,17 @@ protected function errorWithScript( $req ) {
...
+		} else {
+			corsCallback(false);
```
Comment (mgol) [inline test/data/mock.php]: It should be:
```suggestion
			echo 'corsCallback( false )';
```

File: test/data/mock.php [hunk 19]
```diff
@@ -229,6 +229,17 @@ protected function errorWithScript( $req ) {
 		}
 	}
 
+	protected function cors($req) {
```
Comment (mgol) [inline test/data/mock.php]:
```suggestion
	protected function cors( $req ) {
```

**Comment by mgol** [comment]
I figured out the issue. Karma runs on `http://localhost:9876` so a replacement of `localhost` to `127.0.0.1` worked fine there. However, on my local PHP setup I'm relying on virtual hosts so the page runs on `http://jquery.l` so the replacement failed and the loaded script was, in effect, using the same domain. Apparently, in Chrome & Firefox that works but Safari 13 & Edge 18 just don't execute the script.

Ultimately, we cannot rely on the URL where the tests are loaded to be `localhost`. On our TestSwarm, for example, tests are loaded via URLs like:
http://builds.jenkins.jquery.com/jquery/ff2819911da6cbbed5ee42c35d695240f06e65e3/test/index.html
Replacing it via an IP won't work.

We need to remove the replacement of `localhost` to `127.0.0.1` from the test. The mere inclusion of the `crossorigin` attribute makes the script loaded with the `Origin` header so the current test checking whether `corsCallback` was invoked with `true` would fail without current source changes; that may serve as some kind of a check. Unfortunately, if we cut out the `Access-Control-Allow-Methods` & `Access-Control-Allow-Origin` the test still passes; I'm not sure if there's a way around that, though, unfortunately.

File: test/data/mock.php (hunk 17 above)
Comment (mgol) [reply test/data/mock.php]: Hmm, actually, this mistake was why the PHP tests completely failed for me... With these changes they work even in Safari.

Taking that & what I wrote in:
https://github.com/jquery/jquery/pull/4563#issuecomment-575833262
into account, perhaps the current check is fine, we should just add a comment above the `localhost` to `127.0.0.1` replacement saying that full cross-origin verification is only done when tests run from `localhost`, e.g. when run via Karma (as we can depend on `localhost` resolving to `127.0.0.1`) and that in other cases we only do a more limited check that the Origin header is sent.

File: test/unit/ajax.js [hunk 20]
```diff
@@ -2840,6 +2840,32 @@ if ( typeof window.ArrayBuffer === "undefined" || typeof new XMLHttpRequest().re
...
 
+	QUnit.test( "jQuery.append with crossOrigin attribute", function( assert ) {
+		var done = assert.async();
+		assert.expect( 1 );
+
+		var timeout = null;
+
+		Globals.register( "corsCallback" );
+		window.corsCallback = function( response ) {
+			assert.ok( response === true, "Cors" );
```
Comment (mgol) [inline test/unit/ajax.js]: This could be more descriptive:
```suggestion
			assert.ok( response === true, "Callback fired" );
```

File: test/unit/ajax.js (hunk 11 above)
Comment (mgol) [inline test/unit/ajax.js]: A bit more descriptive:
```suggestion
			assert.ok( false, "Callback should have fired" );
```

**Comment by gaohuia** [comment]
Thanks, I learnt a lot in this PR.

**Comment by RiZKiT** [comment]
Regarding Firefox, if you request a script with `crossorigin="anonymous"` but do not provide a `Access-Control-Allow-Origin`header, the script will not be executed, on purpose.

**Comment by gaohuia** [comment]
@mgol Can you help to review it?

**Review by mgol — CHANGES_REQUESTED** [review]
See latest comments.

File: src/manipulation/_evalUrl.js [hunk 21]
```diff
@@ -10,6 +10,7 @@ jQuery._evalUrl = function( url, options, doc ) {
 		cache: true,
 		async: false,
 		global: false,
+		scriptAttrs: options.crossOrigin ? { "crossOrigin": options.crossOrigin } : null,
```
[the remark this answers is not in the document]
Comment (mgol) [reply src/manipulation/_evalUrl.js:13]: The decision from the meeting is to treat it as a breaking change. This means we'll release it in jQuery 4.0 but not before, unless we figure out a backwards-compatible way to land this there.

File: test/middleware-mockserver.js [hunk 22]
```diff
@@ -243,6 +243,15 @@ var mocks = {
 		} else {
 			resp.end( "QUnit.assert.ok( false, \"Mock return erroneously executed\" );" );
 		}
+	},
+	cors: function( req, resp ) {
+		if ( typeof req.headers[ "origin" ] != "undefined" ) {
+			resp.setHeader( "Access-Control-Allow-Methods", "GET" );
```
[the remark this answers is not in the document]
Comment (gaohuia) [reply test/middleware-mockserver.js:249]: Should be good. Will remove it.

File: test/middleware-mockserver.js [hunk 23]
```diff
@@ -243,6 +243,15 @@ var mocks = {
...
+			resp.setHeader( "Access-Control-Allow-Origin", req.headers[ "origin" ] );
```
[the remark this answers is not in the document]
Comment (gaohuia) [reply test/middleware-mockserver.js:250]: I think so.

File: test/middleware-mockserver.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (mgol) [reply test/middleware-mockserver.js:249]: @gaohuia Let's resolve threads only when a change is pushed to make it easier for us to track what's left to do.

**Comment by gaohuia** [comment]
@mgol can you tell me how to run the karma tests with the real mock.php (I mean run it with php)? I added the same logic to "mock.php", but just not sure if it works fine.

**Comment by mgol** [comment]
@gaohuia PHP tests are not run via Karma. To have all tests passing, you need to have Apache or nginx configured with PHP and open `/test/?hidepassed` in a browser, tests will run.

If this is too much, you can run a quick static PHP server in the jQuery directory:
```
php -S 0.0.0.0:7001
```
and then open http://localhost:7001/test/?hidepassed. Note that with this basic setup some tests will fail as PHP can only handle one request at a time and some tests depend on being able to run requests in parallel. But it might be enough for your purposes.

**Comment by gaohuia** [comment]
Thanks @mgol , i have nginx+fpm locally.  the mock.php seems working good.

**Review by mgol — APPROVED** [review]
Looks good to me! Just one suggestion.

In my testing, the new test succeeds on BrowserStack in IE 11 in Windows 10 but not in Windows 7. I'm not sure why, maybe the Windows 7 copy is not up to date. We'll worry about that when/if we start testing on Karma in IE.

Thanks for following up with our code review remarks for this past months!

One note, @gaohuia: we're treating this as a potentially breaking change as code using the `crossorigin` attribute on same-domain scripts would now switch to using the script tag injection method which is asynchronous. Because of that, for now we'll only land it on the `master` branch which is the future jQuery 4.0. We can consider backporting this change to the `3.x-stable` branch if this incompatibility is resolved, maybe by only passing `scriptAttrs` in `jQuery._evalUrl` for cross-domain scripts? We need to be careful here, though, so at least for now it's 4.0-only.

**Comment by mgol** [comment]
Landed, thanks!
